### PR TITLE
MPNet implementation

### PR DESCRIPTION
**Pull request by mehmetcanay**

### CODE REVIEW

File: index/mapping.py [hunk 1]
```diff
@@ -2,16 +2,19 @@
 import numpy as np
 
 from index.embedding import EmbeddingModel
-from index.db.model import Terminology, Mapping, Concept, Variable
```
Comment (tiadams) [inline index/mapping.py:5]: Here might be an issue - the model file was here moved to the db package. Weird that this does not throw an error in the parser test or python package though, I will have to take a look at that.

Could you adjust this?

File: index/embedding.py [hunk 2]
```diff
@@ -28,19 +27,42 @@ def get_embedding(self, text: str, model="text-embedding-ada-002"):
                 return None
             if isinstance(text, str):
                 text = text.replace("\n", " ")
-            return openai.Embedding.create(input=[text], model=model)['data'][0]['embedding']
+            return openai.Embedding.create(input=[text], model=model)["data"][0][
```
Comment (tiadams) [inline index/embedding.py:30]: Not really a problem syntax wise but I am wondering about this formatting choice

File: index/visualisation.py [hunk 3]
```diff
@@ -7,14 +7,15 @@
...
+import plotly.express as px
 
 from index.conf import COLORS_AD, COLORS_PD
 from index.mapping import MappingTable
 
 
 class PlotSide(Enum):
-    LEFT = 1,
-    RIGHT = 2,
+    LEFT = (1,)
```
Comment (tiadams) [inline index/visualisation.py:17]: Why did you add brackets here?

File: index/visualisation.py [hunk 4]
```diff
@@ -30,25 +31,45 @@ def get_cohort_specific_color_code(cohort_name: str):
...
-            "Fuzzy": acc_fuzzy}
+def enrichment_plot(
+    acc_gpt,
+    acc_mpnet,
+    acc_fuzzy,
+    title,
+    save_plot=False,
+    save_dir="resources/results/plots",
+):
+    if (
```
Comment (tiadams) [inline index/visualisation.py:46]: Could be done a bit shorter ;)

if  **not** len(A) == len(B) == len(C) do stuff

File: index/visualisation.py [hunk 5]
```diff
@@ -61,71 +82,222 @@ def concat_embeddings(tables1: [MappingTable], tables2: [MappingTable]):
...
     for table1, table2 in zip(tables1_cleaned, tables2_cleaned):
-        table1.joined_mapping_table.dropna(subset=['embedding', 'description'], inplace=True)
-        table2.joined_mapping_table.dropna(subset=['embedding', 'description'], inplace=True)
-    vectors_tables1 = np.concatenate([table.get_embeddings_numpy() for table in tables1_cleaned])
-    vectors_tables2 = np.concatenate([table.get_embeddings_numpy() for table in tables2_cleaned])
-    descriptions_table1 = np.concatenate([table.joined_mapping_table["description"] for table in tables1_cleaned])
-    descriptions_table2 = np.concatenate([table.joined_mapping_table["description"] for table in tables2_cleaned])
-    boundaries1 = np.array([table.joined_mapping_table["embedding"].index.size for table in tables1_cleaned])
-    boundaries2 = np.array([table.joined_mapping_table["embedding"].index.size for table in tables2_cleaned])
+        table1.joined_mapping_table.dropna(
```
Comment (tiadams) [inline index/visualisation.py:85]: Personally I like to format on one line if possible, but this is fine too

File: index/visualisation.py [hunk 6]
```diff
@@ -61,71 +82,222 @@ def concat_embeddings(tables1: [MappingTable], tables2: [MappingTable]):
...
+    dist1_gpt: pd.DataFrame,
+    dist1_mpnet: pd.DataFrame,
+    dist2_fuzzy: pd.DataFrame,
+    dist2_gpt: pd.DataFrame,
+    dist2_mpnet: pd.DataFrame,
+    title: str,
+    label1: str,
+    label2: str,
+):
+    if not all(
```
Comment (tiadams) [inline index/visualisation.py:136]: Nicely done programming defensively here and checking for exceptions

File: tests/test_evaluation.py [hunk 7]
```diff
@@ -12,62 +17,148 @@ class Test(TestCase):
...
+        self.assertEqual(3 / 5, acc)
 
     def test_match_closest_description_fuzzy(self):
         mapping_table1 = MappingTable(self.mapping_source, self.data_dictionary_source)
         mapping_table2 = MappingTable(self.mapping_source, self.data_dictionary_source)
         # make the second mapping table shorter to test the case where there are more descriptions in the first
-        mapping_table2.joined_mapping_table = mapping_table2.joined_mapping_table.iloc[:-2]
-        result = match_closest_descriptions(mapping_table1, mapping_table2,
-                                            matching_method=MatchingMethod.FUZZY_STRING_MATCHING)
+        mapping_table2.joined_mapping_table = mapping_table2.joined_mapping_table.iloc[
```
Comment (tiadams) [inline tests/test_evaluation.py:105]: This formatting looks counter intuitive to me

File: tests/test_visualisation.py [hunk 8]
```diff
@@ -98,15 +113,52 @@ def test_scatter_plot_all_cohorts(self):
...
 
     def test_enrichment_plot(self):
         acc_gpt = [0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0, 1.0, 1.0, 1.0]
+        acc_mpnet = [0.3, 0.4, 0.5, 0.6, 0.8, 0.9, 1.0, 1.0, 1.0, 1.0]
         acc_fuzzy = [0.2, 0.3, 0.4, 0.5, 0.8, 0.9, 1.0, 1.0, 1.0, 1.0]
         title = "Test"
-        enrichment_plot(acc_gpt, acc_fuzzy, title, save_plot=False)
+        enrichment_plot(acc_gpt, acc_mpnet, acc_fuzzy, title, save_plot=False)
+
+    def test_bar_chart_average_acc_two_distributions(self):
```
Comment (tiadams) [inline tests/test_visualisation.py:135]: You did a great job at exception handling exceptions in your implementation -  best practice would here to also test for the opposite case; meaning you check for each exception case in a separate test function if your expected exception is raised for a wrong function input.

Python uses the [assert.AssertRaises](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaises) function.

File: index/evaluation.py [hunk 9]
```diff
@@ -3,17 +3,24 @@
...
+from scipy.spatial import distance
+from sklearn.metrics.pairwise import cosine_distances
 
 from index.mapping import MappingTable
 
 
 class MatchingMethod(Enum):
-    EUCLIDEAN_EMBEDDING_DISTANCE = 1,
-    FUZZY_STRING_MATCHING = 2
+    EUCLIDEAN_EMBEDDING_DISTANCE = (1,)
```
Comment (tiadams) [inline index/evaluation.py:13]: Why did you add brackets here?

File: index/evaluation.py [hunk 10]
```diff
@@ -51,23 +67,46 @@ def enrichment_analysis(source_table: MappingTable, target_table: MappingTable,
...
+                distances.append(
+                    100 - fuzz.ratio(source_table_description, target_table_description)
+                )
             else:
-                raise NotImplementedError("Specified matching method is not implemented!")
-        min_distance_indices = np.argsort(np.array(distances))[:max_cumulative_match_rank]
+                raise NotImplementedError(
+                    "Specified matching method is not implemented!"
+                )
+        min_distance_indices = np.argsort(np.array(distances))[
```
Comment (tiadams) [inline index/evaluation.py:92]: formatting error?

File: index/evaluation.py [hunk 11]
```diff
@@ -177,3 +271,69 @@ def score_mappings(matches: pd.DataFrame) -> float:
     matches = matches[matches["target_concept_label"].notnull()]
     accuracy = matches["correct"].sum() / len(matches)
     return accuracy
+
+
+def evaluate(
```
Comment (tiadams) [inline index/evaluation.py:276]: Definitely made sense to move this here instead

File: index/evaluation.py [hunk 12]
```diff
@@ -177,3 +271,69 @@ def score_mappings(matches: pd.DataFrame) -> float:
...
+        for idx, source in enumerate(datasets):
+            acc_gpt = []
+            acc_fuzzy = []
+            for idy, target in enumerate(datasets):
+                map_gpt = match_closest_descriptions(source, target)
+                map_fuzzy = match_closest_descriptions(
+                    source, target, matching_method=MatchingMethod.FUZZY_STRING_MATCHING
+                )
+                if target == "jadni":
+                    print("check")
```
Comment (tiadams) [inline index/evaluation.py:296]: What is this check for? :D

File: index/evaluation.py [hunk 13]
```diff
@@ -177,3 +271,69 @@ def score_mappings(matches: pd.DataFrame) -> float:
...
+                acc_gpt.append(round(score_mappings(map_gpt), 2))
+                acc_fuzzy.append(round(score_mappings(map_fuzzy), 2))
+            data_gpt[labels[idx]] = acc_gpt
+            data_fuzzy[labels[idx]] = acc_fuzzy
+        # transpose to have from -> to | row -> column like in the paper
+        gpt = pd.DataFrame(data_gpt, index=labels).T
+        fuzzy = pd.DataFrame(data_fuzzy, index=labels).T
+        return gpt, fuzzy
+
+    elif model == "mpnet":
```
Comment (tiadams) [inline index/evaluation.py:317]: You could probably move code out of each if clause to a higher common level to have less code duplication

File: index/evaluation.py [hunk 14]
```diff
@@ -177,3 +271,69 @@ def score_mappings(matches: pd.DataFrame) -> float:
...
+                    map_mpnet.to_excel(
+                        results_root_dir
+                        + "/mpnet_"
+                        + f"{labels[idx]}_to_{labels[idy]}.xlsx"
+                    )
+                acc_mpnet.append(round(score_mappings(map_mpnet), 2))
+            data_mpnet[labels[idx]] = acc_mpnet
+        # transpose to have from -> to | row -> column like in the paper
+        mpnet = pd.DataFrame(data_mpnet, index=labels).T
+        return mpnet
```
Comment (tiadams) [inline index/evaluation.py:339]: Having a function return a different number of values for different arguments is not a good idea, since it makes the behaviour somewhat less predictable to the user.

Better solution would be here to have it return either gpt/fuzzy/mpnet results (since you already have a model paprameter) .

**Review by tiadams — CHANGES_REQUESTED** [review]
Good ob on the implementations, just some minor comments and suggestions if you want to take a look.

File: index/evaluation.py (hunk 12 above)
Comment (mehmetcanay) [reply index/evaluation.py:296]: It was there in the original code, so I just left it :D

File: index/embedding.py (hunk 2 above)
Comment (mehmetcanay) [reply index/embedding.py:30]: I'm using Black Formatter on VScode and it automatically formats on save. I will try to disable it.

File: index/visualisation.py (hunk 3 above)
Comment (mehmetcanay) [reply index/visualisation.py:17]: It's the Black Formatter again, I will fix it.
